crates/component-echo-http/src/lib.rs

```diff
@@ -0,0 +1,90 @@
+wit_bindgen::generate!({
+    world: "component"
+});
+
+/// Implementation of the 'component' world in wit/world.wit will hang off of this struct
+struct Component;
+
+use crate::exports::wasi::http::incoming_handler::Guest;
+use crate::wasi::http::types::{
+    Fields, IncomingBody, IncomingRequest, Method, OutgoingBody, OutgoingResponse, ResponseOutparam,
+};
+use crate::wasi::io::streams::StreamError;
+
+/// Maximum bytes to read at a time from the incoming request body
+/// this value is chosen somewhat arbitrarily, and is not a limit for bytes read,
+/// but is instead the amount of bytes to be read *at once*
+const MAX_READ_BYTES: u32 = 2048;
+
+impl Guest for Component {
+    fn handle(request: IncomingRequest, outparam: ResponseOutparam) {
+        // Pull together information from the request
+        let method: String = match request.method() {
+            Method::Get => "GET".into(),
+            Method::Post => "POST".into(),
+            Method::Patch => "PATCH".into(),
+            Method::Put => "PUT".into(),
+            Method::Delete => "DELETE".into(),
+            Method::Options => "OPTIONS".into(),
+            Method::Head => "HEAD".into(),
+            Method::Connect => "CONNECT".into(),
+            Method::Trace => "TRACE".into(),
+            Method::Other(m) => m,
+        };
+
+        // Read the body
+        let Ok(incoming_req_body) = request.consume() else {
+            panic!("failed to consume incoming request body");
+        };
+        let Ok(incoming_req_body_stream) = incoming_req_body.stream() else {
+            panic!("failed to build stream from incoming request body");
+        };
+        let mut body_buf = Vec::<u8>::with_capacity(MAX_READ_BYTES as usize);
+        loop {
+            match incoming_req_body_stream.read(MAX_READ_BYTES as u64) {
+                Ok(bytes) => body_buf.extend(bytes),
+                Err(StreamError::Closed) => {
+                    body_buf.shrink_to_fit();
+                    break;
+                }
+                Err(e) => {
+                    panic!("failed to read bytes: {e}");
+                }
+            }
+        }
+        IncomingBody::finish(incoming_req_body);
+
+        // Build the information that will be returned
+        let path_with_query = request
+            .path_with_query()
+            .map(String::from)
+            .unwrap_or_else(|| "/".into());
+        let (path, query) = path_with_query.split_once('?').unwrap_or_else(|| ("/", ""));
+        let response_json = serde_json::json!({
+                    "method": method,
+                    "path": path,
+                    "query_string": query,
+                    "body": body_buf,
+        });
```

Comment:
Is there a reason this had to be JSON?
Why not set the original request paths as response headers and stream the body back as-is?

Reply:
From what I understand, the point of this example (the original `echo` example) is primarily as demo ware, so you can *see* the request (and maybe pipe it to `jq`). 

A similar project is  https://github.com/traefik/whoami. 

I'm fine with changing it, but it's all up to what this component is intended to be used as. @thomastaylor312 @brooksmtownsend @ricochet any comments here, as people who very often do demos?

Reply:
Why not just print text to stdout like `curl -v` or https://github.com/traefik/whoami you mentioned would do?
Then we can trivially just stream the request back.